mqtt/mqtt-broker/src/session.rs

```diff
@@ -458,10 +487,32 @@ impl SessionState {
     }
 
     fn allowed_to_send(&self) -> bool {
+        if self.config.max_inflight_messages() == 0 {
+            return true;
+        }
+
         let num_inflight = self.waiting_to_be_acked.len()
             + self.waiting_to_be_acked_qos0.len()
             + self.waiting_to_be_completed.len();
-        num_inflight < MAX_INFLIGHT_MESSAGES
+        num_inflight < self.config.max_inflight_messages()
+    }
+
+    fn add_to_queue(&mut self, publication: proto::Publication) {
+        if self.config.max_queued_messages() > 0
+            && self.waiting_to_be_sent.len() >= self.config.max_queued_messages()
+        {
+            match self.config.when_full() {
+                QueueFullAction::DropNew => {
+                    // do nothing
+                }
+                QueueFullAction::DropOld => {
+                    let _ = self.waiting_to_be_sent.pop_front();
+                    self.waiting_to_be_sent.push_back(publication);
+                }
+            }
+        } else {
+            self.waiting_to_be_sent.push_back(publication);
```

Comment:
When `self.config.max_queued_messages()` == 0 it will always push messages into the queue. Is that what we want?

Reply:
I assume mosquitto works this way, but I'm not sure we need to do the same. If user wants to opt-out from `max_queued_messages` we have Option for this case. `Option<NonZeroUsize>` ?
https://doc.rust-lang.org/std/num/struct.NonZeroUsize.html

Reply:
Yes, in our case 0 in the config.json would mean "unlimited" queue. But I also agree with using `NonZeroUsize` internally. As discussed offline, let's make that change in the next PR to keep this one shorter.